### PR TITLE
Update CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by vtraag**
I went over our existing CONTRIBUTING.md, and I thought it could use some updating. In particular, I thought it would be good to mention that people can also contribute to the project in other ways than reporting bugs and opening pull request. In particular, I also wanted to clarify to people that help on the support forum and on improving the documentation is helpful.

One other suggestion that I now make is that people can go over existing issues to see if they can help out. Some issues on documentation may be good starting points if people want to contribute, but don't necessarily want to delve intro coding. Some other possibilities would be to just go over some open issues that are already open for a long time and see if they are still relevant. 

I was thinking that it might perhaps be good to have a label for something like "replication needed", so that people can help out by trying to replicate some problem? In addition, we might perhaps clean up a bit our labelling. We now have "help wanted", "PR welcome", perhaps they can simply be combined. But I also wonder: do they actually signify something useful?

In addition, I made some of the language is bit friendlier (or what I would perceive as friendlier) and a bit more welcoming.

This may need some polishing still, feel free to make suggestions!

### CODE REVIEW

**Review by szhorvat — APPROVED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,35 +1,41 @@
...
-they should reciprocate that respect in addressing your issue or assessing
-patches and features.
-
+Thank you for being interested in contributing to `igraph`! We need the help of
+volunteers to keep the package going, so every help is welcome. You can help out
+the project in several different ways. This repository only hosts the C code of
+the `igraph` project. Even if you are not so experienced with C, you can
+contribute in a number of ways.
+
+You can contribute to `igraph` in various ways:
```
Comment (iosonofabio) [inline CONTRIBUTING.md:9]: Good to say this, but perhaps not 3 times in 5 lines?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -79,18 +85,21 @@ Example:
 <a name="features"></a>
 ## Feature requests
 
-Feature requests are welcome. But take a moment to find out whether your idea
-fits with the scope and aims of the project. It's up to *you* to make a strong
-case to convince the project's developers of the merits of this feature. Please
-provide as much detail and context as possible.
-
+Feature requests are always welcome. But take a moment to find out whether your
```
Comment (iosonofabio) [inline CONTRIBUTING.md:88]: Feature requests are always welcome. First, take a moment to find out...

**Comment by iosonofabio** [comment]
Thanks @vtraag, overall in favour. Just suggested a few minor wording things.

I'd be happy with merging "help welcome" into "PR welcome". To me, that means that we think the issue is appropriate for an external user to touch - as opposed to core issues that we don't really want to delegate outside. We could have a "core team" label to signify the opposite: we don't want people to mess with it. We could also start something like an expected difficulty ladder, e.g. "easy", "moderate", "hard"?

Shall we also migrate the wiki pages onto igraph.org? That's be a good excuse to improve the website and unify information.

**Comment by ntamas** [comment]
>  was thinking that it might perhaps be good to have a label for something like "replication needed", so that people can help out by trying to replicate some problem?

Yes, that's a good idea, we can do that.

> We now have "help wanted", "PR welcome", perhaps they can simply be combined.

Sure, we can do that. Many of the labels are not used / obsolete now, in fact some of these labels were migrated from Launchpad when we've had igraph's bug tracker there, way before the Github days. (We even used GNU arch as the version control system at one point, _after_ migrating from SVN, _before_ migrating to Git).

**Comment by ntamas** [comment]
> Shall we also migrate the wiki pages onto igraph.org?

Actually, I like it that we can easily edit the wiki without jumping through hoops with the `igraph.org` repo so I'd rather leave it there.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,35 +1,41 @@
...
-Please take a moment to review this document in order to make the contribution
-process easy and effective for everyone involved.
-
-Following these guidelines helps to communicate that you respect the time of
-the developers managing and developing this open source project. In return,
-they should reciprocate that respect in addressing your issue or assessing
-patches and features.
-
+Thank you for being interested in contributing to `igraph`! We need the help of
+volunteers to keep the package going, so every help is welcome. You can help out
```
Comment (szhorvat) [inline CONTRIBUTING.md:4]: ... so every little bit is welcome ...

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,35 +1,41 @@
...
 
-* Please **do not** use the issue tracker for personal support requests (use
-  our [igraph support forum](https://igraph.discourse.group)).
-
-* Please **do not** derail or troll issues. Keep the discussion on topic and
-  respect the opinions of others.
-
-Please also take a look at our [tips on writing igraph code](#tips) before
-getting your hands dirty.
+- Do you have a support question? Please use
```
Comment (szhorvat) [inline CONTRIBUTING.md:27]: "Do you have questions?"

Any _question_ should go to the forum, no need to say "support" question.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -132,25 +140,25 @@ project:
    git checkout -b <topic-branch-name>
    ```
 
-4. Commit your changes in logical chunks. Please adhere to these [git commit
-   message guidelines](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)
-   or your code is unlikely be merged into the main project. Use Git's
-   [interactive rebase](https://help.github.com/articles/interactive-rebase)
-   feature to tidy up your commits before making them public.
+4. Please commit your changes in logical chunks, and try to provide clear commit
```
Comment (szhorvat) [inline CONTRIBUTING.md:143]: I wonder if we need this sentence. Most beginner PRs will be squash-and-merged. It's easier for people to add a lot of increment commits without having to worry about keeping it tidy. Then it's all flattened into a single commit in the end. This is what we do as well.

What do you think?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -132,25 +140,25 @@ project:
...
-6. Locally merge (or rebase) the upstream development branch into your topic branch:
+6. Make sure that your PR is based off the latest code and locally merge (or
+   rebase) the upstream development branch into your topic branch:
 
    ```bash
    git pull [--rebase] upstream <dev-branch>
    ```
 
-7. Push your topic branch up to your fork:
+7. If necessary, you can use git's
```
Comment (szhorvat) [inline CONTRIBUTING.md:158]: Same comment as for point 4.

Rebase is dangerous, and one can accidentally ruin one's work with it. I have. Do we want this in the beginner instructions?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -159,17 +167,47 @@ project:
...
-**IMPORTANT**: By submitting a pull request, you agree to allow the project owner to
-license your work under the same license as that used by the project.
+**IMPORTANT**: By submitting a pull request, you agree to allow the project
+owner to license your work under the same license as that used by the project,
+see also [Legal Stuff](#legal).
 
 <a name="branching"></a>
 ### Branching
 
-`igraph` is committed to [semantic versioning](https://semver.org/). We are currently still in the development release (0.x), which in principle is a mark that the public API is not yet stable. Regardless, we try to maintain semantic versioning also for the development releases. We do so as follows. Any released minor version (0.x.z) will be API backwards-compatible with any previous release of the *same* minor version (0.x.y, with y < z). This means that *if* there is an API incompatible change, we will increase the minor version. For example, release 0.8.1 is API backwards-compatible with release 0.8.0, while release 0.9.0 is API incompatible with version 0.8.1. Note that this only concerns the *public* API, internal functions may change also within a minor version.
```
Comment (szhorvat) [inline CONTRIBUTING.md:168]: The Branching section is much more detailed than the rest is not directly relevant to contributing. While this is good information, does it belong into this document?  Should we perhaps significantly shorten it and write it in the form of easy-to-follow instructions about whether to target `master` or `develop`?

**Comment by szhorvat** [comment]
I was already happy with it as it was, I just made a few small suggestions. Up to you if you act on them.

File: CONTRIBUTING.md (hunk 5 above)
Comment (ntamas) [reply CONTRIBUTING.md:143]: I think it does not hurt the review process if the commits are neat and tidy; I tend to go through a PR commit-by-commit to understand how the contribution was built up from the ground up, although sometimes I simply give up if the dead ends are also there.

File: CONTRIBUTING.md (hunk 6 above)
Comment (ntamas) [reply CONTRIBUTING.md:158]: Do you mean only the _interactive_ rebasing or the `--rebase` option in general from point 6?

I agree that interactive rebasing is advanced stuff; even I use it only in rare occasions. However, rebasing when pulling is probably useful as the alternative is merging, in which case you need to deal with merge conflicts. And I think there is always an easy way out of `git pull --rebase upstream` by running `git rebase --abort`.

Comment (szhorvat) [reply CONTRIBUTING.md:158]: Yes, I mean the interactive rebasing. I have messed up my work when doing interactive rebasing, encountering conflicts and failing to clean them up correctly. All I wanted was a tidy commit history (point 4). I think we should not encourage that in this document.

Comment (szhorvat) [reply CONTRIBUTING.md:158]: > However, rebasing when pulling is probably useful

Yes, I have no problem with that. That is useful to mention.

File: CONTRIBUTING.md (hunk 7 above)
Comment (ntamas) [reply CONTRIBUTING.md:168]: I decided to keep this for the time being so I can merge the PR; we can always re-write it or simplify it in a later step.

**Review by ntamas — APPROVED** [review]

**Comment by vtraag** [comment]
Thanks for the additional improvements! And, yes, better indeed to remove interactive rebasing.

**Pull request merged**